[숫자 야구] 정재우 미션 제출합니다.

holyPigeon opened the pull request:
## Intro

안녕하세요, 프리코스 여러분!

리뷰하시기 편하도록 `패키지 구조`와 `리뷰가 필요한 부분`를 적어보았습니다. 이전까지 나름 자바를 많이 안다고 생각했지만 이번 미션을 거치면서 머리 끝까지 탈탈 털리게 되었는데요. 

막판 가서 정신없이 리팩토링을 한지라 리뷰하시면서 뭔가 `굳이 여기에 이런 코드가 쓰여야 하나?` 또는 `왜 이렇게 뭐가 뒤죽박죽이지?` 식의 의문이 드실 수도 있습니다. ~~정상입니다ㅠㅠ~~

어떤 의견이든 남겨주신다면 도움이 많이 될 것 같습니다! 서로 리뷰하면서 성장하는 시간 가졌으면 좋겠습니다🔥

## 패키지 구조

```
.
└── baseball
    ├── Application.java --- 전체 어플리케이션 실행을 담당하는 클래스
    ├── application 
    │   └── NumberService.java --- 야구 게임의 숫자와 관련된 로직 처리를 담당하는 클래스
    ├── constant
    │   ├── ErrorMessage.java --- 에러 메시지 관련 enum
    │   ├── HintMessage.java --- 힌트 메시지 관련 enum
    │   ├── RestartOption.java --- 게임 재시작 및 종료 옵션 관련 enum
    │   └── SystemMessage.java --- 시스템 메시지 관련 enum
    ├── controller
    │   └── GameController.java --- 게임 진행을 총괄하는 컨트롤러 클래스
    ├── domain
    │   ├── Hint.java --- 숫자 야구 게임의 힌트 관련 도메인
    │   └── Player.java --- 숫자 야구 게임을 플레이하는 플레이어 관련 도메인
    ├── util
    │   ├── NumberUtil.java --- 문자열과 리스트 간 변환, 그리고 검증 기능을 포함한 유틸 클래스
    │   └── validateCondition
    │       ├── LengthCondition.java --- 입력값 길이 검증 관련 enum
    │       └── RangeCondition.java --- 입력값 범위 검증 관련 enum
    └── view
        ├── InputView.java --- 사용자 입력을 받는 뷰 클래스
        └── OutputView.java --- 메시지를 출력하는 뷰 클래스
```

## 리뷰가 필요한 부분

- MVC 패턴의 역할이 잘 나눠져 있나요?

- 플레이어와 컴퓨터의 숫자를 비교하여 힌트를 도출해내는 기능에서 비효율적인 로직은 없었나요?

- 패키지 구조는 보기 쉽게 잘 나눠져 있나요?

- 디미터의 법칙이 잘 적용되었나요?

- 특정 로직들이 다른 클래스, 또는 다른 계층에 위치했으면 좋겠다 싶은 것들이 있나요?

- enum 활용에서 더 개선할 점이 있을까요?

## 기타

이 외에도 코드를 보면서 피드백해주실 부분, 또는 이해가 안 가거나 궁금한 부분이 있으시면 **정말 아무거나라도 좋으니** 의견 남겨주시면 감사하겠습니다!!!

holyPigeon committed:
docs: 기능 구현 목록 정리

holyPigeon committed:
docs: 기능 구현 목록 수정

holyPigeon committed:
feat: Player 클래스 생성

게임을 플레이하는 플레이어에 관한 객체

holyPigeon committed:
feat: Player 클래스 필드 변수 구현

플레이어의 숫자를 나타내는 변수

holyPigeon committed:
feat: NumberService 클래스 생성

플레이어의 숫자와 관련된 기능을 담담하는 객체

holyPigeon committed:
feat: 컴퓨터의 숫자를 랜덤하게 정하는 메소드 구현

holyPigeon committed:
feat: ViewService 클래스 생성

입력과 출력을 담당하는 객체

holyPigeon committed:
feat: 사용자 숫자 입력 메시지 출력 메소드 구현

holyPigeon committed:
docs: 기능 구현 목록 수정

holyPigeon committed:
feat: 사용자 숫자 입력 메소드 구현

holyPigeon committed:
feat: NumberUtil 클래스 생성

플레이어 숫자와 관련된 유틸리티 기능을 담당하는 클래스

holyPigeon committed:
feat: 입력된 문자열이 비어있지 않은지 검증하는 메소드 구현

holyPigeon committed:
feat: 입력된 문자열의 길이가 3인지 검증하는 메소드 구현

holyPigeon committed:
feat: 입력된 문자열이 정수인지 검증하는 메소드 구현

holyPigeon committed:
feat: 입력된 문자열이 양의 정수인지 검증하는 메소드 구현

holyPigeon committed:
docs: 기능 구현 목록 수정

holyPigeon committed:
feat: 문자열 타입의 숫자를 리스트로 변환하는 메소드 구현

holyPigeon committed:
feat: 문자열 타입의 숫자를 리스트로 변환하는 메소드 구현

holyPigeon committed:
feat: 입력된 정수들의 범위가 1부터 9인지 검증하는 메소드 구현

holyPigeon committed:
feat: 입력된 정수들의 범위가 1부터 9인지 검증하는 메소드 구현

holyPigeon committed:
feat: 입력값을 전체적으로 검증하는 메소드 구현

holyPigeon committed:
feat: NumberUtil 클래스 protected 생성자 적용

같은 패키지 외에서는 NumberUtil의 인스턴스를 생성을 막고, 정적 메소드를 호출할 수 있도록 함.

holyPigeon committed:
feat: 숫자 비교에 따라 힌트를 결정하는 메소드 구현

holyPigeon committed:
feat: 숫자 비교에 따라 힌트를 결정하는 메소드 구현

holyPigeon committed:
feat: 플레이어가 입력한 숫자의 위치를 비교하는 메소드 구현

holyPigeon committed:
feat: 플레이어가 입력한 각 숫자가 컴퓨터의 숫자에 있는지 확인하는 메소드 구현

holyPigeon committed:
feat: Hint 클래스 생성

숫자 비교에 따른 힌트와 관련된 클래스

holyPigeon committed:
feat: Hint 클래스 필드 변수 구현

strike 개수와 ball 개수를 나타내는 변수

holyPigeon committed:
feat: Hint 클래스 생성자 구현

Hint 클래스가 처음 생성될 때, strike과 ball의 개수를 모두 0으로 초기화 한다.

holyPigeon committed:
feat: strike과 ball의 개수를 증가시키는 메소드 구현

holyPigeon committed:
feat: strike과 ball이 각각 존재하는지 확인하는 메소드 구현

holyPigeon committed:
feat: strike이 3개인지 확인하는 메소드 구현

holyPigeon committed:
feat: strike와 ball의 각 개수에 따라 문자열을 반환하는 메소드 구현

holyPigeon committed:
feat: 조건에 따라 문자열을 이어붙여 힌트 문자열을 반환하는 메소드 구현

holyPigeon committed:
feat: 게임 재시작 메시지를 출력하는 메소드 구현

holyPigeon committed:
feat: 게임 종료 메시지를 출력하는 메소드 구현

holyPigeon committed:
feat: 힌트 메시지를 출력하는 메소드 구현

holyPigeon committed:
feat: GameController 클래스 생성

전체 게임 진행을 관리하는 클래스

holyPigeon committed:
feat: GameController 클래스 의존성 주입 로직 구현

NumberService와 ViewService의 의존성을 주입받는다.

holyPigeon committed:
feat: Player 클래스 getter, setter 메소드 구현

holyPigeon committed:
feat: 컴퓨터의 숫자를 세팅하는 메소드 구현

holyPigeon committed:
feat: 플레이어의 숫자를 세팅하는 메소드 구현

holyPigeon committed:
feat: 야구 게임 진행 메소드 구현

holyPigeon committed:
feat: 야구 게임 시작 메소드 구현

holyPigeon committed:
feat: 야구 게임 어플리케이션 실행 로직 구현

holyPigeon committed:
docs: 리팩토링 목록 작성

holyPigeon committed:
docs: 기존 과제 설명 및 요구 사항 제거

holyPigeon committed:
refector: 숫자 비교에 따른 힌트 설정 로직 최적화

holyPigeon committed:
refector: 일부 for문을 stream 메소드를 사용하여 간소화

holyPigeon committed:
fix: 정수 범위 검증 로직에서 발생하는 에러 수정

해당 로직이 ASCII 값을 기반으로 숫자를 검증하도록 수정

holyPigeon committed:
docs: README.md와 docs/README.md의 혼용 수정

holyPigeon committed:
refector: Hint 클래스 관련 힌트 메시지 목록을 Enum으로 분리

holyPigeon committed:
refector: 입력값 검증 시 발생하는 예외의 목록을 Enum으로 분리

holyPigeon committed:
refector: strike 및 ball 개수에 따른 상태 반환 메소드 최적화

holyPigeon committed:
refector: 입력값 검증 시의 검증 조건들을 Enum으로 분리

holyPigeon committed:
refector: View 계층에서 출력하는 시스템 메시지의 목록을 Enum으로 분리

holyPigeon committed:
refector: ViewService의 기능을 InputView와 OutputView로 각각 분리

holyPigeon committed:
fix: 정수 범위 검증 로직에서 발생하는 에러 수정

해당 로직이 ASCII 값을 정수 값으로 변환하여 검증하도록 수정

holyPigeon committed:
feat: 문자열을 정수 타입의 리스트로 변환하는 유틸 메소드 구현

holyPigeon committed:
feat: Player 클래스의 setter를 제거하고 생성 시에만 필드 변수값을 정할 수 있도록 수정

holyPigeon committed:
fix: 리스트를 문자열로 변환하는 유틸 메소드가 제대로 동작하지 않는 오류 수정

holyPigeon committed:
refector: 싱글톤 패턴 적용

holyPigeon committed:
refector: ViewService 클래스 삭제

holyPigeon committed:
refector: GameController의 메소드명을 명시적으로 변경

holyPigeon committed:
refector: 힌트 생성 시의 숫자 비교 로직 메소드 분리

holyPigeon committed:
refector: 힌트 생성 메소드명을 명시적으로 변경

holyPigeon committed:
refector: Player 객체를 불변 객체로 만들기 위한 로직 구현

1. final을 통한 재할당 방지
2. 방어적 복사 적용을 통해 생성 시점의 변수 조작 방지
3. unmodifiableList 적용을 통해 getter 사용으로 인한 변수 조작 방지

holyPigeon committed:
refector: 게임 종료 및 재시작 enum 추가 구현

holyPigeon committed:
refector: 게임 실행 조건을 가독성 향상을 위해 메소드로 분리

holyPigeon committed:
refector: 싱글톤 패턴 적용에 따른 기본 생성자 접근 제한

h-beeen commented on src/main/java/baseball/application/NumberService.java:
```diff
@@ -0,0 +1,61 @@
+
+public class NumberService {
+
+    private static final NumberService instance = new NumberService();
+
+    private NumberService() {
+
+    }
+
+    public static NumberService getInstance() {
+        return instance;
+    }
```
NumberService 자체를 static으로 생성한 이유가 있었을까요!!?
NumberService 자체를 싱글톤으로 활용하실 목적이라고 판단되는데,
하위 메소드를 static으로 선언해서 생성자 없이 활용하는 방안도 직관적이었을 것 같습니다!

holyPigeon replied:
싱글톤 패턴을 적용하려고 한 코드가 맞습니다! 처음 적용시에는 static 키워드가 정확히 어떤 역할을 하는지 잘 몰랐는데 질문에 답하려다 보니 **전역적으로 NumberService 인스턴스를 하나만 생성**하기 위한 역할이 있다는 것을 알게 되었습니다...!

아무래도 NumberService 클래스에 대한 접근을 하나의 인스턴스로만 관리하고자 하는 생각이 있어서 말씀해주신 방법을 적용하지는 않겠지만, 덕분에 왜 제가 이런 방식으로 코드를 짜야 하는지에 대해서 더 깊게 생각해본 계기가 된 것 같습니다.

조언 감사합니다!!!

Laniel88 replied:
저도 윗분 댓글을 읽고 생각해보았는데, NumberService 자체는 상태(인스턴스 변수)를 갖지 않아 메소드를 static으로 선언해서 사용하는 것이 더 직관적일 것 같다는 생각이 듭니다.

h-beeen commented on src/main/java/baseball/application/NumberService.java:
```diff
@@ -0,0 +1,61 @@
+
+    public List<Integer> getRandomNumber() {
+        Set<Integer> numberSet = new HashSet<>();
+        while (numberSet.size() < 3) {
```
매직넘버 리터럴이 사용되었어요!
상수를 포장하거나, enum으로 configuration을 세팅하면 더 좋을 것 같아요!

holyPigeon replied:
Enum을 다 만들어놓고 적용을 안 시켰네요... 날카로운 조언 감사합니다!

h-beeen commented on src/main/java/baseball/application/NumberService.java:
```diff
@@ -0,0 +1,61 @@
+    public List<Integer> getRandomNumber() {
+        Set<Integer> numberSet = new HashSet<>();
+        while (numberSet.size() < 3) {
+            numberSet.add(Randoms.pickNumberInRange(1, 9));
```
이부분도 매직넘버 리터럴이 사용되었네요!

holyPigeon replied:
매직넘버 전부 없애버리겠습니다...!

h-beeen commented on src/main/java/baseball/constant/ErrorMessage.java:
```diff
@@ -0,0 +1,23 @@
+
+public enum ErrorMessage {
+
+    INPUT_EMPTY("입력값은 비어있지 않아야 합니다."),
+    INPUT_LENGTH_INVALID("입력값은 3자리이어야 합니다."),
+    INPUT_NUMBER_NOT_INTEGER("입력값은 정수이어야 합니다."),
+    INPUT_NUMBER_NOT_POSITIVE_INTEGER("입력값은 양의 정수이어야 합니다."),
+    INPUT_NUMBER_NOT_IN_RANGE("입력되는 각 숫자는 1 이상 9 이하의 정수로 이루어져야 합니다."),
+    INPUT_NUMBER_NOT_UNIQUE("입력되는 각 숫자는 서로 다른 숫자이어야 합니다.");
```
에러메세지를 enum으로 처리한 부분 아주 인상적입니다!

holyPigeon replied:
감사합니다!!!

h-beeen commented on src/main/java/baseball/constant/HintMessage.java:
```diff
@@ -0,0 +1,19 @@
+
+public enum HintMessage {
+
+    ONLY_BALL("%d볼"),
+    ONLY_STRIKE("%d스트라이크"),
+    NOTHING("낫싱");
```
서식지정자 포맷을 이용해, n볼, n스트라이크로 처리하신 부분도 좋아요!

holyPigeon replied:
잘하는 분 코드를 좀 참고해봤는데 상당히 좋은 방법인 것 같습니다👍 감사합니다!

h-beeen commented on src/main/java/baseball/constant/SystemMessage.java:
```diff
@@ -0,0 +1,20 @@
+    TYPE_NUMBER("숫자를 입력해주세요 : "),
+    GAME_END("3개의 숫자를 모두 맞히셨습니다! 게임 종료"),
+    GAME_RESTART("게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요."),
+    EXIT_FLAG("1"),
+    RESTART_FLAG("2");
```
해당 부분은 SystemMessage라는 변수 명에 적합하지 않은 친구들인 것 같아요.
그리고 FLAG는 T/F의 의미가 강해서, EXIT_FLAG 대신 `EXIT_VALUE`나 그냥 `EXIT`로 선언해도 좋았을 것 같아요!

holyPigeon replied:
1. 얘기를 듣고 보니 SystemMessage 말고 RestartOption? 등의 이름으로 바꾸는 게 더 명시적일 것 같네요! 
2. FLAG라는 단어가 통상적으로 그렇게 쓰이는지 처음 알았네요...! 말씀하신대로 하면 헷갈릴 여지가 없을 것 같습니다! 

좋은 의견 감사합니다!!!

h-beeen commented on src/main/java/baseball/controller/GameController.java:
```diff
@@ -0,0 +1,52 @@
+
+public class GameController {
+
+    private final NumberService numberService = NumberService.getInstance();
+    private final InputView inputView = InputView.getInstance();
+    private final OutputView outputView = OutputView.getInstance();
```
아까 언급드렸던 ServiceLayer에서, 단순히 Instance를 멤버변수 없이 생성해서 리턴하는 로직이라면,
해당 부분을 전부 삭제하고, 메소드를 static method로 호출해도 좋았을 것 같아요.

holyPigeon replied:
좋은 의견 감사합니다!!!

h-beeen commented on src/main/java/baseball/domain/Hint.java:
```diff
@@ -0,0 +1,71 @@
+        return isStrikeExist() && isBallExist();
+    }
+
+    public boolean isStrikeExist() {
+        return strike > 0;
+    }
+
+    public boolean isBallExist() {
+        return ball > 0;
+    }
+
+    public boolean isThreeStrike() {
+        return strike == 3;
+    }
```
매직넘버 리터럴이 사용되었습니다!!

holyPigeon replied:
매직넘버를 우주에서 지워버리겠습니다

ㅋㅋㅋㅋ

h-beeen commented on src/main/java/baseball/domain/Hint.java:
```diff
@@ -0,0 +1,71 @@
+        return join(" ", resultString);
+    }
+
+    public String getStrikeString() {
+        return format(ONLY_STRIKE.getMessage(), strike);
+    }
+
+    public String getBallString() {
+        return format(ONLY_BALL.getMessage(), ball);
```
format을 사용해 아까 enum에서 선언한 서식지정자를 합쳐서 사용하는 부분이네요!
인상적입니다 :)

holyPigeon replied:
`형식 지정자를 사용한 Enum`과 `format`의 궁합이 좀 야무진 것 같습니다. 감사합니다!!

h-beeen commented on src/main/java/baseball/domain/Player.java:
```diff
@@ -0,0 +1,22 @@
+        this.numberList = new ArrayList<>(numbers);
+    }
+
+    public List<Integer> getNumberList() {
+        return Collections.unmodifiableList(numberList);
+    }
```
일급컬렉션을 수정할 수 없도록 참조해서 복사하도록 설계하셨군요! 인상적이에요!

holyPigeon replied:
어떻게 하면 getter를 제거할 수 있을까를 계속해서 고민해보았는데, 초반에 설계를 잘못해서 결국에는 다 갈아엎지는 못하고 `unmofiableList()`라는 기능을 찾아서 사용하게 되었습니다...! 다음에는 웬만하면 getter 자체를 사용 안 하도록 노력해보려고 합니다.

칭찬 감사합니다!!!

h-beeen commented on src/main/java/baseball/view/OutputView.java:
```diff
@@ -0,0 +1,42 @@
+    public void printHintMessage(Hint hint) {
+        System.out.println(hint.toString());
+
+        if (hint.isThreeStrike()) {
+            printGameEndMessage();
+        }
```
View는 비즈니스 로직을 몰라야 한다고 생각해요!
이 부분은 컨트롤러에서 처리하고, View에서는 단순히 print하는 방법으로 설계해도 좋았을 것 같습니다 :)

holyPigeon replied:
만들 때는 이 정도는 괜찮겠지?라고 생각했는데, View가 다른 클래스에 완벽하게 의존하지 않으려면 저런 로직들도 다 빼야겠네요...! 

좋은 의견 감사합니다!!! 반영해볼게요

h-beeen reviewed (approved):
깊은 고민 끝에 작성한 코드라는 것을 느낄 수 있었습니다!
매직넘버 리터럴 사용, View Layer의 책임과 관련된 몇몇 리뷰 남겨드렸습니다:)
좋은 코드 잘 보고, 많이 배우고갑니다.
감사합니다!

holyPigeon committed:
refector: 랜덤 숫자 생성 로직에 관련 enum 적용

holyPigeon committed:
refector: NumberService 클래스의 일부 메소드에 적용된 불필요한 static 키워드 제거

holyPigeon committed:
refector: 게임 재시작 옵션에 관한 SystemMessage enum 필드 제거

holyPigeon committed:
refector: 게임 재시작 옵션에 관한 RestartOption enum 생성

holyPigeon committed:
refector: 게임 재시작 판단 로직에 관련 enum 적용

holyPigeon committed:
refector: 힌트 출력 메시지의 비즈니스 로직을 Controller 계층으로 이동

holyPigeon committed:
refector: 게임 종료 및 재시작 메시지 출력 메소드의 접근 제한자 변경

printGameEndMessage(): Controller에 사용하기 위해 public으로 변경
printRestartMessage(): printGameEndMessage의 부속 메소드로만 쓰이기 때문에 private으로 변경

holyPigeon committed:
feat: 입력에 쓰인 Scanner를 닫는 기능 구현

holyPigeon committed:
feat: 게임 종료 후 입력을 종료하는 로직 반영

jinny-l commented on src/main/java/baseball/util/NumberUtil.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    public static String convertListToString(List<Integer> numberList) {
+        return numberList.stream()
+                .map(String::valueOf)
+                .collect(Collectors.joining(""));
```
`String.join()`을 사용하면 스트림을 돌지 않고 문자열을 합칠 수 있어요!

holyPigeon replied:
사실 저 부분은 저도 `String.join()` 기능을 써보려고 했는데, `List<String>` 타입만 지원하고 `List<Integer>`는 들어가지 않더라구요ㅠㅠ

한 줄로 끝날 수 있는 로직인데 스트림을 써서 3줄로 가버리니까 마음이 정말 아팠습니다... 혹시 List<Integer> 타입의 변수를 곧바로 String으로 바꾸는 더 간결한 코드가 있는지 궁금하네요...!

jinny-l commented on src/main/java/baseball/util/NumberUtil.java:
```diff
@@ -0,0 +1,80 @@
+        }
+    }
+
+    protected static void validateSign(String number) {
+        if (number.charAt(0) == '-') {
+            throw new IllegalArgumentException(INPUT_NUMBER_NOT_POSITIVE_INTEGER.getMessage());
+        }
+    }
```
음수 여부를 검증하기 위한 메서드라고 생각이 드는데요,
개인적으로는 파싱해서 부호를 검증하기보다는 문자열을 숫자로 변환 후 범위 검증을 하는 것이 가독성이 더 좋다고 느껴져요!

이거는 개인적인 의견이니 참고만 해주세요 🙂

holyPigeon replied:
생각해보니 정수로 변환한 후에 0보다 작은지에 대해 비교해보면 되는 거였네요! 괜히 어렵게 생각한 것 같습니다 😂

좋은 의견 감사합니다!!!

jinny-l commented on src/main/java/baseball/view/InputView.java:
```diff
@@ -0,0 +1,26 @@
+        return readLine();
+    }
+
+    public void readClose() {
+        close();
```
Scanner close까지 챙겨주셨군요.
꼼꼼한 처리가 인상 깊습니다!

holyPigeon replied:
저도 처음엔 몰랐는데 다른 분 코드를 보다가 제공된 `Console` API에 close 메소드가 있다는 것을 알게 되어서 한 번 써봤습니다...!

칭찬 감사합니다!

jinny-l commented on docs/README.md:
```diff
@@ -0,0 +1,60 @@
+# 미션 - 숫자 야구
+
```
리드미가 깔끔합니당 👍

jinny-l commented on src/main/java/baseball/application/NumberService.java:
```diff
@@ -0,0 +1,62 @@
+        return new ArrayList<>(numberSet);
+    }
+
+    public Hint generateHint(Player computer, Player player) {
+        List<Integer> computerNumber = computer.getNumberList();
+        List<Integer> playerNumber = player.getNumberList();
+        Hint hint = new Hint();
+
+        compareNumber(computerNumber, playerNumber, hint);
+
+        return hint;
+    }
```
`NumberService`에서 `Player` 숫자를 꺼내와서 직접 비교하게 구현하셨군요!
`Player` 객체 내부에 비교 로직을 구현하게 되면 `getter`를 사용하지 않고도 똑같은 로직을 구현할 수 있을 것 같아요~

직접적으로 객체의 상태를 외부에 제공하는 것보다는 외부에서 원하는 결과를 받을 수 있도록 적절한 책임을 객체에 주는 것은 어떨까요?~

holyPigeon replied:
처음에는 **도메인은 최대한 간단하게** 만들고 **서비스에 비즈니스 로직을 몰아넣을 생각**이었는데, 주변에서 getter나 setter가 들어가면 말을 듣고 횝쓸려서 급하게 막 고치다가 나온 결과물인 것 같습니다...ㅠㅠ

말씀해주신대로 객체에 책임을 주려면 저런 비교 로직들도 전부 Player 객체 안에 넣는 게 맞는 것 같습니다!

좋은 조언 감사합니다!!!

jinny-l commented on src/main/java/baseball/domain/Hint.java:
```diff
@@ -0,0 +1,71 @@
+        ball++;
+    }
+
+    public boolean isHintNothing() {
+        return isStrikeExist() && isBallExist();
+    }
```
엇 이거 혹시 제가 이해한게 맞을까요??
낫싱 상황을 판단하는 메서드인데 스트라이크와 볼이 모두 존재할 때 true를 리턴하는 것처럼 보이는데
반대로 리턴해야 하는게 아닌가 해서요

holyPigeon replied:
아이고... 말씀해주신 게 맞습니다.  저도 지금 봤네요ㅠㅠ 반대로 하는 게 맞는 것 같습니다.

테스트를 도대체 어떻게 통과한 건지 봤더니 만들어놓고 저 메소드는 쓰지도 않았었네요 🤣

날카롭게 봐주셔서 감사합니다...!

jinny-l commented on src/main/java/baseball/domain/Hint.java:
```diff
@@ -0,0 +1,71 @@
+        }
+        if (isStrikeExist()) {
+            resultString.add(getStrikeString());
+        }
+        return join(" ", resultString);
```
static import를 하셔서 처음에 join 메서드를 직접 구현하셨나? 생각했어요 ㅎㅎ
혹시 static import를 사용하신 이유가 있을까요??

holyPigeon replied:
아무래도 같은 메서드를 자주 사용하게 될 때 중복된 부분을 없애면서 가독성이 높아지기도 하고, 앞에 `String.`부분을 빼더라도 의미는 전달될 거라 생각해서 static import를 사용하게 되었습니다!

사실은 약간 습관의 부분도 있는 것 같네요...! 저 부분은 큰 생각 없이 static import를 사용했는데, 말씀하신 것처럼 해당 메소드가 어디로부터 온 것인지 알아보기 어려울 수도 있겠다는 생각이 듭니다!

아무래도 여태 습관처럼 아무데나 쓰기 보다는 정말 필요한 순간을 생각해서 써보는 것이 좋겠네요! 좋은 질문 감사합니다 😀

jinny-l reviewed (commented):
읽는동안 코드에서 깊은 고민이 느껴졌습니다 ㅎㅎ
1주차 미션 너무너무 고생 많으셨어요!

잘 짜신 부분에 따봉 남기려고 했는데, 앞서 리뷰 남겨주신 분이 다 언급해주셨더라구요 ㅎㅎ
코드 잘봤습니당 🙂

chs98412 commented on src/main/java/baseball/constant/HintMessage.java:
```diff
@@ -0,0 +1,19 @@
+
+public enum HintMessage {
+
+    ONLY_BALL("%d볼"),
+    ONLY_STRIKE("%d스트라이크"),
+    NOTHING("낫싱");
```
이 부분 배워갑니다!

chs98412 commented on src/main/java/baseball/constant/SystemMessage.java:
```diff
@@ -0,0 +1,18 @@
+
+public enum SystemMessage {
+
+    TYPE_NUMBER("숫자를 입력해주세요 : "),
+    GAME_END("3개의 숫자를 모두 맞히셨습니다! 게임 종료"),
+    GAME_RESTART("게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요.");
+
+    private final String message;
+
+    SystemMessage(String message) {
+        this.message = message;
+    }
+
+    public String getMessage() {
+        return message;
```
enum을 하나로 다 묶은게 아니라 용도별로 따로 구성한게 인상적입니다!

Laniel88 commented on src/main/java/baseball/constant/ErrorMessage.java:
```diff
@@ -0,0 +1,23 @@
+package baseball.constant;
+
+public enum ErrorMessage {
```
지금까지 `public static final`로만 상수를 모으는 코드를 보았었는데
메세지 문자열을 enum으로 처리하는 코드는 처음 봐서 인상적이었습니다. (enum에 생성자를 사용할 수 있는지조차 몰랐습니다.)
코드 길이가 줄어드는 것 외에 어떤 장점이 더 있다고 생각하시나요?

holyPigeon replied:
`static final`을 이용한 상수 처리와 비교하면 아무래도 enum에서는 상수들을 이용하여 메소드를 정의할 수 있다보니, 좀 더 활용도가 높다는 장점이 있는 것 같습니다!

Laniel88 commented on src/main/java/baseball/util/NumberUtil.java:
```diff
@@ -0,0 +1,80 @@
+
+public class NumberUtil {
+
+    protected NumberUtil() {
```
`NumberUtil`의 생성자를 `private`으로 써도 괜찮아보이는데 혹시 `protected`로 쓰신 이유가 있을까요?

holyPigeon replied:
해당 부분은 스프링 데이터 JPA 공부를 하면서 자주 짰던 코드가 습관으로 나온 것 같습니다...! 

본래의 의도는 접근을 완전히 제한하지 않고 `protected`로 설정하여 같은 패키지 내에서는 접근을 허가하도록 하는 것인데요. 아무래도 유틸 클래스 특성상, 생성자를 생성할 일이 거의 없을테니 여기서는 `private`이 어울리는 게 맞는 것 같습니다!

캐치하지 못한 부분 알려주셔서 감사합니다 😄

Laniel88 commented on src/main/java/baseball/util/validateCondition/LengthCondition.java:
```diff
@@ -0,0 +1,20 @@
+package baseball.util.validateCondition;
+
+public enum LengthCondition {
```
각 유효성을 enum에서 분리해서 다루는 점이 간결하고 좋아보입니다!

holyPigeon replied:
저도 여러 코드를 참고하다가 발견한 부분인데요, enum을 이용하면 이렇게 검증 조건과 메소드를 한 곳에 두는 게 가능하더라구요! 

enum의 다양한 활용 방안에 대해 계속해서 연구해봐야할 것 같습니다 🙂

Laniel88 reviewed (commented):
1주차 수고하셨습니다. 제가 배울게 더 많은 코드였네요.
